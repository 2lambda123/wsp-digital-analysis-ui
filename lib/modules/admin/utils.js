// @flow
import {connect} from 'react-redux'
import {createStructuredSelector} from 'reselect'

import * as actions from './actions'
import * as select from './selectors'

export function fullyConnect (Component: any) {
  return connect(
    createStructuredSelector(select),
    actions
  )(Component)
}
