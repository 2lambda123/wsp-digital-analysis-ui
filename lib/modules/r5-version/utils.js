//
import {VERSION_PARSE_REGEX} from './constants'

/**
 * Convert an r5 version to a number for comparing
 */
export function versionToNumber(version) {
  const versionMatch = (version || '').match(VERSION_PARSE_REGEX)
  if (versionMatch) {
    const [, major, minor, patch, commit] = versionMatch

    // NB Assumes no major, minor, patch, or commit version exceeds 100
    return (
      parseInt(major) * 10000 +
      parseInt(minor) * 100 +
      parseInt(patch) +
      parseInt(commit || 0) / 100
    )
  }

  return 0
}
