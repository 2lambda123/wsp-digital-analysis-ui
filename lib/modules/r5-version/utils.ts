import {MINIMUM_R5_VERSION, VERSION_PARSE_REGEX} from './constants'

/**
 * Convert an r5 version to a number for comparing
 */
export function versionToNumber(version: string): number {
  const versionMatch = (version || '').match(VERSION_PARSE_REGEX)
  if (versionMatch) {
    const [, major, minor, patch, commit] = versionMatch

    // NB Assumes no major, minor, patch, or commit version exceeds 100
    return (
      parseInt(major) * 10000 +
      parseInt(minor) * 100 +
      parseInt(patch) +
      parseInt(commit || '0') / 100
    )
  }

  return 0
}

/**
 * Check if the r5 version is older than a given minimum.
 */
export function workerVersionInRange(
  version: string,
  min: string = MINIMUM_R5_VERSION,
  max = 'v100'
): boolean {
  const versionNumber = versionToNumber(version)
  return (
    versionNumber >= versionToNumber(min) &&
    versionNumber <= versionToNumber(max)
  )
}
