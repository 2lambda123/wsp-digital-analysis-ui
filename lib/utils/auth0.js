// @flow
import message from '@conveyal/woonerf/message'
import get from 'lodash/get'

const CLIENT_ID = process.env.AUTH0_CLIENT_ID
const DOMAIN = process.env.AUTH0_DOMAIN

let LogRocket = {identify () {}}
if (process.env.LOGROCKET) {
  LogRocket = require('logrocket')
  LogRocket.init(process.env.LOGROCKET)
}

// If client & domain exist, then require authentication
export const authIsRequired: boolean = !!(CLIENT_ID && DOMAIN)

// Default Conveyal logo & color
const logo = 'https://d2f1n6ed3ipuic.cloudfront.net/conveyal-128x128.png'
const primaryColor = '#2389c9'

// The default Auth0 scope to request
const scope = 'openid email profile analyst offline_access'

// Default Auth0 response type
const responseType = 'code id_token token'

/**
 * Initialize in a function call so that the Auth0 lock code isn't executed
 * until needed. When Auth0 code is "required" it saves the origin. This is a
 * problem when a user goes to another part of the site (ex: `/regions/create`)
 * and we want to `checkSession` before allowing them to continue. Each "origin"
 * must be saved in the Auth0 Dashboard (which is impossible due to IDs). So
 * instead we redirect to `/login`, `checkSession`, and redirect back to the
 * original path given.
 */
function initializeLock () {
  const Auth0Lock = require('auth0-lock').default
  return new Auth0Lock(CLIENT_ID, DOMAIN, {
    auth: {
      params: {
        responseType,
        scope
      },
      redirect: false
    },
    allowSignUp: false,
    theme: {
      logo,
      primaryColor
    },
    closable: false,
    languageDictionary: {
      title: message('authentication.logIn')
    }
  })
}

/**
 * Initialize Auth0, check the user session, and prompt for a login if needed
 */
export function login (setUser: (any) => void) {
  const lock = initializeLock()

  function handleAuthResult (error, authResult) {
    if (error) {
      return lock.show({
        flashMessage: {
          type: 'error',
          text: error.error_description
        }
      })
    }
    if (!authResult) return lock.show()

    // Flatten the user object
    const user = {
      ...authResult,
      ...authResult.idTokenPayload // contains email, name, nickname, stored metadata
    }

    // Check for the access group and show error if it does not exist
    const accessGroup = get(user, 'analyst.group')
    if (typeof accessGroup !== 'string') {
      return lock.show({
        flashMessage: {
          type: 'error',
          text: `User ${user.email} does not have access to this application.`
        }
      })
    }

    // Log user activities
    LogRocket.identify(user.userId || user.user_id, {
      accessGroup,
      email: user.email
    })

    // AuthResult OK, pass result on. idTokenPayload
    setUser(user)
    lock.hide()
  }

  // Set the default handler for auth errors
  lock.on('authorization_error', handleAuthResult)
  lock.on('authenticated', (authResult) => handleAuthResult(null, authResult))
  lock.on('unrecoverable_error', handleAuthResult)

  // Check the session by default
  lock.checkSession({
    responseType,
    scope,
    timeout: '2000'
  }, handleAuthResult)
}

/**
 * Log out and reload the page.
 */
export function logout () {
  const lock = initializeLock()

  lock.logout({
    clientID: CLIENT_ID,
    returnTo: window.location.origin
  })
}
