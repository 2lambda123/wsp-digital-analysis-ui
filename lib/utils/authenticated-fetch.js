import dbg from 'debug'
import fetch from 'isomorphic-fetch'
import store from '../store' // TODO: don't import store from anywhere but the root/bootstrap

const debug = dbg('scenario-editor:authenticated-fetch')

export default async function authenticatedFetch (url, options = {}) {
  debug(`${options.method || 'get'} ${url.split('?')[0]}`)
  const headers = options.headers || {}
  const state = store.getState()
  if (state.user && state.user.id_token) {
    headers.Authorization = `bearer ${state.user.id_token}`
  }
  const response = await fetch(url, Object.assign({}, options, {headers}))
  debug(`${options.method || 'get'} ${url.split('?')[0]} is ${response.status} ${response.statusText}`)
  return response
}

export function parseJSON (res) {
  if (res.ok) {
    return res.json()
  } else {
    window.alert(res.statusText)
    throw new Error(res.statusText)
  }
}
