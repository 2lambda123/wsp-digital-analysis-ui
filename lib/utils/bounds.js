// @flow
import message from '@conveyal/woonerf/message'
import {LatLng, LatLngBounds} from 'leaflet'

import reprojectCoordinates from './reproject-coordinates'

import type {Bounds} from '../types'

const ANTIMERIDIAN_WARNING = 'We do not currently support bounds that cross the antimeridian.'

export function toLatLngBounds (bounds: Bounds) {
  return new LatLngBounds([
    [bounds.north, bounds.west],
    [bounds.south, bounds.east]
  ])
}

/**
 * NB: We do not currently support crossing the antimeridian.
 */
export function fromLatLngBounds (bounds: LatLngBounds) {
  // Don't change the original bounds object
  let b = new LatLngBounds()
  b.extend(bounds)

  if (b.getWest() < -180) {
    b = new LatLngBounds(
      new LatLng(b.getSouth(), -180),
      b.getNorthEast()
    )

    window.alert(message('antimeridian', ANTIMERIDIAN_WARNING))
  }

  if (b.getEast() > 180) {
    b = new LatLngBounds(
      new LatLng(b.getSouth(), 180),
      b.getNorthEast()
    )
    window.alert(message('antimeridian', ANTIMERIDIAN_WARNING))
  }

  const bboxString = b.toBBoxString()
  return {
    north: b.getNorth(),
    south: b.getSouth(),
    east: b.getEast(),
    west: b.getWest(),
    toString: () => bboxString
  }
}

/**
 * Reproject bounds to the grid size that we use.
 */
export function reprojectBounds (bounds: LatLngBounds) {
  return new LatLngBounds(
    reprojectCoordinates(bounds.getSouthWest()),
    reprojectCoordinates(bounds.getNorthEast())
  )
}
