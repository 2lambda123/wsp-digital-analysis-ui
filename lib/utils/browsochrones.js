/** Handle interactions with Browsochrones and the cluster */

import Browsochrones from 'browsochrones'
import {crc32} from 'crc'
import fetch from 'isomorphic-fetch'
import dbg from 'debug'
const debug = dbg('scenario-editor:browsochrones')
import uuid from 'uuid'

import convertToR5Modification from './convert-to-r5-modification'
import authenticatedFetch from './authenticated-fetch'
import timeout from './timeout'

// TODO get from project
const WORKER_VERSION = 'v1.5.0-78-g5d11cf9'

/** How often to retry 202 response from the server */
const RETRY_TIMEOUT_MILLISECONDS = 10 * 1000

const PROFILE_REQUEST_DEFAULTS = {
  date: (new Date()).toISOString().split('T')[0],
  fromTime: 25200,
  toTime: 32400,
  accessModes: 'WALK',
  directModes: 'WALK',
  egressModes: 'WALK',
  transitModes: 'TRANSIT',
  walkSpeed: 1.3888888888888888,
  bikeSpeed: 4.166666666666667,
  carSpeed: 20,
  streetTime: 90,
  maxWalkTime: 20,
  maxBikeTime: 20,
  maxCarTime: 45,
  minBikeTime: 10,
  minCarTime: 10,
  suboptimalMinutes: 5,
  reachabilityThreshold: 0,
  bikeSafe: 1,
  bikeSlope: 1,
  bikeTime: 1,
  bikeTrafficStress: 4,
  boardingAssumption: 'RANDOM',
  monteCarloDraws: 220
}

// Map from scenario ID to object containing Browsochrones instance, hash of last used scenario
// x, y coords of origin
// TODO LRU cache
let instances = new Map()

// Grids are shared among all Browsochrones instances
// array should contain { name, grid }
let grids = null

/** get a new Browsochrones instance with grids already loaded */
async function newInstance () {
  if (grids == null) throw new Error('Grids are not loaded!')

  let browsochrones = new Browsochrones()

  await grids.map(({ grid, name }) => browsochrones.putGrid(name, grid))

  return {
    // CRC of the scenario, used to detect when stop trees and metadata must be refetched
    scenarioHash: null,
    origin: { x: NaN, y: NaN },
    staticSiteRequest: null,
    browsochrones
  }
}

/** Attempt to fetch a file from the server, retrying on a 202 (Worker starting up) response */
async function retryFetch (url, opts, timeoutMs = RETRY_TIMEOUT_MILLISECONDS) {
  let res = await authenticatedFetch(url, opts)

  if (res.status === 202) {
    debug(`202 response from server, retrying in ${timeoutMs / 1000}s`)
    await timeout(timeoutMs)
    return retryFetch(url, opts, timeoutMs)
  }

  return res
}

/** Set the grids being used, in the format [{name, grid}...] */
export function setGrids (newGrids) {
  if (instances.size > 0) throw new Error('attempt to set grids after Browsochrones instances have been created')
  grids = newGrids
}

/**
 * Get an R5 profile request for a scenario.
 * TODO we need to add a lot more options to this function (date, time, worker version)
 */
function getStaticSiteRequest ({ modifications, bundleId, scenarioUniqueId }) {
  return {
    transportNetworkId: bundleId, // TODO is this necessary?
    workerVersion: WORKER_VERSION,
    bucket: null, // don't save to S3
    jobId: uuid.v4(),
    request: {
      ...PROFILE_REQUEST_DEFAULTS,
      scenario: {
        id: scenarioUniqueId,
        modifications: modifications.map(convertToR5Modification)
      }
    }
  }
}

/** Make sure a scenario is loaded and we are ready to generate isochrones, paths, etc. with Browsochrones */
async function ensureScenarioLoaded ({ scenarioId, bundleId, modifications }) {
  // ensure we have a Browsochrones instance for this scenario id
  if (!instances.has(scenarioId)) {
    debug(`no browsochrones instance for scenario ${scenarioId}, creating one and loading grids`)
    instances.set(scenarioId, await newInstance())
    debug(`created browsochrones instance for scenario ${scenarioId}`)
  }

  let instance = instances.get(scenarioId)

  // hash the modifications to detect any changes that have taken place
  let scenarioHash = crc32(JSON.stringify(modifications))

  // Maybe one day we'll check if anything relevant (the street network) has changed,
  // that would be a performance win
  if (scenarioHash !== instance.scenarioHash) {
    debug(`scenario ${scenarioId} has changed, updating browsochrones`)
    let scenarioUniqueId = `${scenarioId}-${scenarioHash}`
    let request = getStaticSiteRequest({ modifications, bundleId, scenarioUniqueId })
    // TODO don't duplicate these fields
    let { workerVersion, jobId } = request
    let body = {
      request,
      graphId: bundleId,
      workerVersion,
      jobId
    }

    let metadata = await retryFetch('/api/analysis/enqueue/single', {
      method: 'POST',
      body: JSON.stringify({ ...body, type: 'static-metadata' })
    }).then(res => res.json())

    // technically we shouldn't need to retry here as the worker is started once the above request
    // succeeds, but why not.
    let stopTrees = await retryFetch('/api/analysis/enqueue/single', {
      method: 'POST',
      body: JSON.stringify({ ...body, type: 'static-stop-trees' })
    }).then(res => res.arrayBuffer())

    let { browsochrones } = instance

    await Promise.all([
      browsochrones.setQuery(metadata),
      browsochrones.setStopTrees(stopTrees),
      browsochrones.setTransitiveNetwork(metadata.transitive) // TODO shouldn't have to do this.
    ])

    instance.staticSiteRequest = request
    instance.scenarioHash = scenarioHash
    instance.origin = { x: -1, y: -1 }

    debug(`(re-)loaded scenario ${scenarioId}`)
  }
}

/**
 * Ensure that an origin is loaded and a surface has been generated.
 * Assumes that the proper scenario/stop trees have been loaded with ensureScenarioLoaded
 */
async function ensureOriginLoaded ({ scenarioId, x, y }) {
  let instance = instances.get(scenarioId)

  if (instance.origin.x !== x || instance.origin.y !== y) {
    debug(`fetching origin (${x}, ${y}) for scenario ID ${scenarioId}`)

    // TODO don't duplicate these fields
    let { workerVersion, jobId, transportNetworkId } = instance.staticSiteRequest
    let body = {
      request: instance.staticSiteRequest,
      graphId: transportNetworkId,
      workerVersion,
      jobId,
      x,
      y,
      type: 'static'
    }

    // always possible that the worker has died in the meantime, use retryFetch
    let origin = await retryFetch('/api/analysis/enqueue/single', {
      method: 'POST',
      body: JSON.stringify(body)
    }).then(res => res.arrayBuffer())

    await instance.browsochrones.setOrigin(origin, { x, y })
    await instance.browsochrones.generateSurface()
    instance.origin = { x, y }
  }
}

/** Get isochrones and accessibility numbers for a scenario and a given cutoff. Modifications is in scenario-editor format. */
export async function getIsochronesAndAccessibility ({ scenarioId, bundleId, modifications, isochroneCutoff, origin }) {
  await ensureScenarioLoaded({ scenarioId, modifications, bundleId })
  let { browsochrones } = instances.get(scenarioId)
  let { x, y } = browsochrones.latLonToOriginPoint(origin)
  await ensureOriginLoaded({ scenarioId, x, y })
  return {
    isochrone: await browsochrones.getIsochrone(isochroneCutoff)
    // TODO accessibility
  }
}

/** clear all browsochrones instances when changing projects */
export function clear () {
  instances = new Map()
}
