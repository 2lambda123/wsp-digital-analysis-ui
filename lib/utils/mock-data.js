import multi from '@conveyal/woonerf/store/multi'
import promise from '@conveyal/woonerf/store/promise'
import {middleware as fetchMiddleware} from '@conveyal/woonerf/fetch'
import configureStore from 'redux-mock-store'
import thunkMiddleware from 'redux-thunk'

import {ADD_TRIP_PATTERN} from '../constants'

export const makeMockStore = configureStore([fetchMiddleware, multi, promise, thunkMiddleware])

export const mockSegmentGeom = {
  type: 'LineString',
  coordinates: [
    [
      -122.0246,
      36.9707
    ],
    [
      -122.0279,
      37.049
    ],
    [
      -121.9799,
      37.2299
    ],
    [
      -121.9445,
      37.324
    ],
    [
      -121.936,
      37.353
    ],
    [
      -121.924,
      37.365
    ]
  ]
}

export const mockSegment = {
  type: 'Feature',
  properties: {},
  geometry: mockSegmentGeom
}

const mockStop1 = {
  stop_id: '1',
  stop_lat: 36.9707,
  stop_lon: -122.0246,
  stop_name: 'Santa Cruz'
}

const mockStop2 = {
  stop_id: '2',
  stop_lat: 37.049,
  stop_lon: -122.0279,
  stop_name: 'Scotts Valley'
}

export const mockStops = [mockStop1, mockStop2]

export const mockPattern = {
  geometry: mockSegmentGeom,
  stops: mockStops,
  trips: [
    {
      trip_id: 'abcd',
      start_time: 12345,
      trip_short_name: 'The Express',
      trip_headsign: 'To Downtown',
      duration: 1234
    }
  ]
}

export const mockRoute = {
  id: 'route1',
  patterns: [mockPattern]
}

export const mockFeed = {
  id: '1',
  routes: [mockRoute],
  routesById: {
    route1: mockRoute
  },
  stopsById: {
    '1': mockStop1,
    '2': mockStop2
  }
}

export const mockMapState = {
  allowExtend: true,
  extendFromEnd: true,
  followRoad: true,
  state: 'add-trip-pattern',
  modificationId: '1234'
}

export const mockTimetable = {
  name: 'Test timetable',
  segmentSpeeds: [],
  dwellTime: 10,
  monday: true,
  tuesday: true,
  wednesday: true,
  thursday: true,
  friday: true,
  saturday: false,
  sunday: false,
  headwaySecs: 900, // 15 minutes
  startTime: 28800, // 8am
  endTime: 57600, // 4pm
  patternTrips: ['abcd'],
  sourceTrip: 'abcd'
}

export const mockModification = {
  bidirectional: false,
  entries: [],
  feed: mockFeed.id,
  id: '1234',
  name: 'Test Modification',
  routes: ['route1'],
  segments: [],
  showOnMap: false,
  timetables: [mockTimetable],
  trips: ['abcd'],
  type: ADD_TRIP_PATTERN,
  variants: [0]
}

export const mockProject = {
  bounds: {
    north: 39,
    east: -76,
    south: 38,
    west: -77
  },
  description: 'Project description',
  id: '1',
  name: 'Mock Project'
}

export const mockScenario = {
  bundleId: '1',
  feeds: [mockFeed],
  feedsById: {
    [mockFeed.id]: mockFeed
  },
  id: '1',
  name: 'Mock Scenario',
  projectId: '1',
  variants: ['Default']
}

export const mockStores = {
  init: {
    analysis: {
      isFetchingIsochrone: false,
      isochroneCutoff: 60
    },
    mapState: {
      activeTrips: [],
      center: {
        lat: 38.8886,
        lon: -77.043
      },
      components: [],
      zoom: 12
    },
    network: {
      error: null,
      outstandingRequests: 0
    },
    project: {
      currentProject: null,
      projects: [],
      projectsById: {}
    },
    routing: {
      locationBeforeTransitions: null
    },
    scenario: {
      bundles: [],
      bundlesById: {},
      currentBundle: {
        id: '1'
      },
      currentScenario: mockScenario,
      feeds: [mockFeed],
      feedsById: {
        [mockFeed.id]: mockFeed
      },
      modifications: [mockModification],
      modificationsById: {
        [mockModification.id]: mockModification
      },
      modificationsByType: {
        [mockModification.type]: [mockModification]
      },
      scenarios: [],
      scenariosById: {},
      variants: ['Default']
    },
    user: {
      idToken: null
    }
  }
}

export const mockScenarioApplicationError = {
  'modificationId': 'Add Trip Pattern',
  'title': "Error while applying the modification entitled 'Add Trip Pattern'.",
  'messages': [
    'Headway is not greater than zero.'
  ]
}

export const mockProfileRequest = {
  date: '2016-01-16',
  fromTime: 25200,
  toTime: 32400,
  accessModes: 'WALK',
  directModes: 'WALK',
  egressModes: 'WALK',
  transitModes: 'TRANSIT',
  walkSpeed: 1.3888888888888888,
  bikeSpeed: 4.166666666666667,
  carSpeed: 20,
  streetTime: 90,
  maxWalkTime: 20,
  maxBikeTime: 20,
  maxCarTime: 45,
  minBikeTime: 10,
  minCarTime: 10,
  suboptimalMinutes: 5,
  reachabilityThreshold: 0,
  bikeSafe: 1,
  bikeSlope: 1,
  bikeTime: 1,
  bikeTrafficStress: 4,
  monteCarloDraws: 200,
  maxRides: 4
}

export const mockRegionalAnalyses = [
  {
    zoom: 9,
    width: 300,
    height: 300,
    north: 20024,
    west: 20015,
    name: 'ANALYSIS 1',
    grid: 'Jobs_total',
    scenarioId: '1234',
    variant: 0,
    bundleId: '5678',
    cutoffMinutes: 60,
    id: 'abcd',
    request: mockProfileRequest
  },
  {
    zoom: 9,
    width: 300,
    height: 300,
    north: 20024,
    west: 20015,
    name: 'ANALYSIS 2',
    grid: 'Jobs_total',
    scenarioId: '4321',
    variant: 1,
    bundleId: '5678',
    cutoffMinutes: 60,
    id: 'efgh',
    request: mockProfileRequest
  }
]

export const mockGrid = {
  zoom: 9,
  width: 300,
  height: 300,
  north: 20024,
  west: 20015,
  data: [],
  min: 0,
  max: 2000
}

export const mockBookmark = {
  name: 'Bookmark',
  profileRequest: mockProfileRequest,
  isochroneCutoff: 42,
  destinationGrid: 'Workers_total',
  projectId: 'PROJECT'
}
