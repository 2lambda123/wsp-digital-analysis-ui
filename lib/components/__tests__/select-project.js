// @flow
import React from 'react'

import SelectProject from '../select-project'
import {mockWithProvider, mockProject, mockRegion} from '../../utils/mock-data'

describe('Component > SelectProject', () => {
  it('renders correctly', () => {
    const mockProjects = [mockProject]
    const pushFn = jest.fn()
    const {snapshot} = mockWithProvider(<SelectProject
      region={mockRegion}
      projects={mockProjects}
    />)
    expect(snapshot()).toMatchSnapshot()
    expect(pushFn).not.toBeCalled()
  })
})
