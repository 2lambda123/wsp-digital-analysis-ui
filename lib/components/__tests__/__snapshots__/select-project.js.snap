// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component > SelectProject renders correctly 1`] = `
<div
  className="InnerDock"
>
  <div
    className="panel-body"
  >
    <div
      className="WelcomeTitle"
    >
      <p
        style={
          Object {
            "textAlign": "center",
          }
        }
      >
        welcome to
      </p>
      <span
        className="logo"
        to="/"
      >
        conveyal analysis
      </span>
    </div>
    <div
      className="form-group"
    >
      <a
        className="btn btn-success btn-block"
        href="#"
        onClick={[Function]}
        tabIndex={0}
      >
        <i
          className="fa fa-cubes fa-fw "
        />
         
        Set up a new region
      </a>
    </div>
    <div>
      <p
        style={
          Object {
            "textAlign": "center",
          }
        }
      >
        or go to an existing region
      </p>
      <div
        className="list-group"
      >
        <a
          className="list-group-item"
          onClick={[Function]}
          style={Object {}}
          title="Go to region"
        >
          <div>
            <p>
              <i
                className="fa fa-spinner fa-fw fa-spin"
              />
               
              P1
            </p>
            <em />
          </div>
        </a>
        <a
          className="list-group-item"
          onClick={[Function]}
          style={Object {}}
          title="Go to region"
        >
          <div>
            <p>
              <i
                className="fa fa-spinner fa-fw fa-spin"
              />
               
              P2
            </p>
            <em />
          </div>
        </a>
      </div>
    </div>
  </div>
</div>
`;
